### ISSUE TITLE
Is it normal to get large temp files after using mongodb-memory-server?

### ISSUE THREAD

**Issue by edshen17** [nodkz/mongodb-memory-server#530]
## Versions

- NodeJS: 0.0.0
- mongodb-memory-server-*: 0.0.0
- mongodb: 0.0.0
- mongoose: 0.0.0 
- system: 

package: mongo-memory-server 

## What is your question?
After running tests that depend on mongodb-memory-server, I noticed that my temp folder _(C:\Users\e17\AppData\Local\Temp)_ was getting filled with large folders that were 300MB each. They all started with the name "mongo-mem--." I'm not sure what this is, but is this expected behavior? Should I be clearing the temp folder after I am done with my tests? Or is there a way I can automate this?

As for my code, I am not doing anything crazy-- I just connect to MongoMemoryReplSet and perform simple mongodb operations. Here are my replica set settings:
```
{ replSet: { count: 1, storageEngine: 'wiredTiger' } }
```

Thanks,
Edwin.

**Comment by hasezoey** [comment]
it is expected behavior in certain circumstances:
- NodeJS Process fails (internal error)
- Process got killed with signal `SIGKILL`
- The Test Environment fails and the Instance was created *inside a test* (this should only happen in MMS tests (this repository))
- The Test Environment fails to execute `process.on('exit')` listeners added in hooks (like in jest `beforeAll`)
- Explicitly Disabling cleanup with `.stop(false)` (and not having a custom `db_path` specified)
- Using a Custom `db_path` and not running `.cleanup()`

if none of these would match your case, could you provide more information, like in what environment your execute it in (test framework etc) and the code where start / stop is called?

**Comment by edshen17** [comment]
Any updates on this?

**Comment by hasezoey** [comment]
> Any updates on this?

not really, because i am not able to reproduce this locally

> I also tried calling cleanup() after stop(), but got an error:

were you still using version `7.3.6` at that point? when yes, could you try again with a `7.4.0-beta.x` version?
and provide debug output

also, this seems like the error says, using a function wrongly (with the 6.x style in 7.x), could you also provide your startup and teardown code?

**Comment by edshen17** [comment]
I tried running .cleanup() after .stop() and it still does not work. After some testing, it seems like this issue is not related to sessions/transactions at all. This is because I ran my tests without any transaction logic and I still got 300MB folders.

I was wondering if there was a way to examine the content of the temp files so I can pinpoint what model causes this to happen. I tried opening diagnostic.data/metrics.INTERIM but it seems to be encoded. It also seems to be related to using the WiredTiger engine because when I ran my tests without using ReplSet I did not get this issue.

**Comment by hasezoey** [comment]
> I was wondering if there was a way to examine the content of the temp files so I can pinpoint what model causes this to happen

there is not a official way, but it could be easily done by copying the server config from your test startups to the NodeJS REPL (or a dedicated file), and instead of temporary folder, you point them to the existing folders (with instance option `dbPath`)

**Comment by edshen17** [comment]
Hello, after some testing, I think I found out the issue. In my tests, I use a function makeDb() which creates a replset. However, I noticed I was accidentally creating a new replset each time. In my mocha global teardown, it seems like I was stopping only 1 instance rather than all the instances I created. [Here](https://github.com/edshen17/mem-server-demo) is a repo of a basic example.

Running tests/good.test.ts does not cause any issues. However, running tests/bad.test.ts creates 300MB folders in the temp folder. If you check makeDb.ts, you can see that I have mongod outside the makeDb function and I call await mongod() to get the uri. Then in bad.test.ts, in the after hook (line 18), I most likely am not stopping the correct replset instance.

Do you know how I can only create one replset instance and use the same one wherever I call makeDb()? That way, I can call .stop() on one instance in my global teardown.

Thanks!

**Comment by hasezoey** [comment]
> For future reference, maybe you can add a section in the documentation say that you have to call .stop() to cleanup the files when using mocha. Otherwise, I'm really grateful for your time! Thanks again!

yes i will probably add a note about that, but aside from that it seems like mocha just kills the process after all tests run and does not fire / execute `process.on('exit')` events (a hook for each instance gets added by default to call stop (and non-force cleanup))
